// @ant-design/pro-form/lib/layouts/DrawerForm/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DrawerForm = DrawerForm;

require("antd/es/drawer/style");

var _drawer = _interopRequireDefault(require("antd/es/drawer"));

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread2"));

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _react = _interopRequireWildcard(require("react"));

var _useMergedState5 = _interopRequireDefault(require("rc-util/lib/hooks/useMergedState"));

var _reactDom = require("react-dom");

var _omit = _interopRequireDefault(require("omit.js"));

var _BaseForm = require("../../BaseForm");

var _warning = require("rc-util/lib/warning");

var _excluded = ["children", "trigger", "onVisibleChange", "drawerProps", "onFinish", "title", "width"];

function DrawerForm(_ref) {
  var _rest$submitter, _rest$submitter2;

  var children = _ref.children,
      trigger = _ref.trigger,
      onVisibleChange = _ref.onVisibleChange,
      drawerProps = _ref.drawerProps,
      onFinish = _ref.onFinish,
      title = _ref.title,
      width = _ref.width,
      rest = (0, _objectWithoutProperties2.default)(_ref, _excluded);

  var _useMergedState = (0, _useMergedState5.default)(!!rest.visible, {
    value: rest.visible,
    onChange: onVisibleChange
  }),
      _useMergedState2 = (0, _slicedToArray2.default)(_useMergedState, 2),
      visible = _useMergedState2[0],
      setVisible = _useMergedState2[1];

  var footerRef = (0, _react.useRef)(null);
  /** Modal dom 解除渲染之后 */

  var _useMergedState3 = (0, _useMergedState5.default)(false),
      _useMergedState4 = (0, _slicedToArray2.default)(_useMergedState3, 2),
      isDestroy = _useMergedState4[0],
      setIsDestroy = _useMergedState4[1];

  var needRenderForm = (0, _react.useMemo)(function () {
    if (drawerProps === null || drawerProps === void 0 ? void 0 : drawerProps.destroyOnClose) {
      return visible;
    }

    return true;
  }, [drawerProps === null || drawerProps === void 0 ? void 0 : drawerProps.destroyOnClose, visible]);
  (0, _warning.noteOnce)( // eslint-disable-next-line @typescript-eslint/dot-notation
  !rest['footer'] || !(drawerProps === null || drawerProps === void 0 ? void 0 : drawerProps.footer), 'DrawerForm 是一个 ProForm 的特殊布局，如果想自定义按钮，请使用 submit.render 自定义。');
  (0, _react.useEffect)(function () {
    if (visible && rest.visible) {
      onVisibleChange === null || onVisibleChange === void 0 ? void 0 : onVisibleChange(true);
    } // eslint-disable-next-line react-hooks/exhaustive-deps

  }, [visible]);
  /** 设置 trigger 的情况下，懒渲染优化性能；使之可以直接配合表格操作等场景使用 */

  var isFirstRender = (0, _react.useRef)(!(drawerProps === null || drawerProps === void 0 ? void 0 : drawerProps.forceRender));
  /** 如果 destroyOnClose ，重置一下表单 */

  (0, _react.useEffect)(function () {
    if (visible) {
      isFirstRender.current = false;
    }
  }, [drawerProps === null || drawerProps === void 0 ? void 0 : drawerProps.destroyOnClose, visible]);

  var triggerDom = /*#__PURE__*/_react.default.createElement(_react.default.Fragment, {
    key: "trigger"
  }, trigger && /*#__PURE__*/_react.default.cloneElement(trigger, (0, _objectSpread2.default)((0, _objectSpread2.default)({}, trigger.props), {}, {
    onClick: function () {
      var _onClick = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee(e) {
        var _trigger$props, _trigger$props$onClic;

        return _regenerator.default.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                setVisible(!visible);
                (_trigger$props = trigger.props) === null || _trigger$props === void 0 ? void 0 : (_trigger$props$onClic = _trigger$props.onClick) === null || _trigger$props$onClic === void 0 ? void 0 : _trigger$props$onClic.call(_trigger$props, e);

              case 2:
              case "end":
                return _context.stop();
            }
          }
        }, _callee);
      }));

      function onClick(_x) {
        return _onClick.apply(this, arguments);
      }

      return onClick;
    }()
  })));

  (0, _react.useEffect)(function () {
    if (visible) {
      setTimeout(function () {
        setIsDestroy(visible);
      }, 100);
      return;
    }

    setIsDestroy(visible);
  }, [setIsDestroy, visible]);
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_drawer.default, (0, _extends2.default)({
    title: title,
    width: width || 800
  }, drawerProps, {
    visible: visible,
    onClose: function onClose(e) {
      var _drawerProps$onClose;

      setVisible(false);
      drawerProps === null || drawerProps === void 0 ? void 0 : (_drawerProps$onClose = drawerProps.onClose) === null || _drawerProps$onClose === void 0 ? void 0 : _drawerProps$onClose.call(drawerProps, e);
    },
    footer: rest.submitter !== false && /*#__PURE__*/_react.default.createElement("div", {
      ref: footerRef,
      style: {
        display: 'flex',
        justifyContent: 'flex-end'
      }
    })
  }), needRenderForm && /*#__PURE__*/_react.default.createElement(_BaseForm.BaseForm, (0, _extends2.default)({
    formComponentType: "DrawerForm",
    layout: "vertical"
  }, (0, _omit.default)(rest, ['visible']), {
    submitter: rest.submitter === false ? false : (0, _objectSpread2.default)((0, _objectSpread2.default)({}, rest.submitter), {}, {
      searchConfig: (0, _objectSpread2.default)({
        submitText: '确认',
        resetText: '取消'
      }, (_rest$submitter = rest.submitter) === null || _rest$submitter === void 0 ? void 0 : _rest$submitter.searchConfig),
      resetButtonProps: (0, _objectSpread2.default)({
        preventDefault: true,
        onClick: function onClick(e) {
          var _drawerProps$onClose2;

          setVisible(false);
          drawerProps === null || drawerProps === void 0 ? void 0 : (_drawerProps$onClose2 = drawerProps.onClose) === null || _drawerProps$onClose2 === void 0 ? void 0 : _drawerProps$onClose2.call(drawerProps, e);
        }
      }, (_rest$submitter2 = rest.submitter) === null || _rest$submitter2 === void 0 ? void 0 : _rest$submitter2.resetButtonProps)
    }),
    onFinish: /*#__PURE__*/function () {
      var _ref2 = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee2(values) {
        var success;
        return _regenerator.default.wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                if (onFinish) {
                  _context2.next = 2;
                  break;
                }

                return _context2.abrupt("return");

              case 2:
                _context2.next = 4;
                return onFinish(values);

              case 4:
                success = _context2.sent;

                if (success) {
                  setVisible(false);
                }

              case 6:
              case "end":
                return _context2.stop();
            }
          }
        }, _callee2);
      }));

      return function (_x2) {
        return _ref2.apply(this, arguments);
      };
    }(),
    contentRender: function contentRender(item, submitter) {
      // 未配置自定义提交按钮，则直接将提交按钮渲染到内容区
      if (rest.submitter === false) {
        return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, item, submitter);
      } // 如果用户配置了自定义的提交按钮，那么我们等到弹框/抽屉底部区域渲染成功后再将自定义按钮渲染过去


      if (footerRef.current && isDestroy && submitter) {
        return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, item, /*#__PURE__*/(0, _reactDom.createPortal)(submitter, footerRef.current));
      }

      return item;
    }
  }), children)), triggerDom);
}
